feat: Remove docker_py usage in favor of custom client that uses Docker CLI

angela-ko opened the pull request:
#### Relevant issue or PR
Fixes #10 
Enables #8

#### Description of changes
- Created a new `docker_client.py` file that contains a `CLIDockerClient` class that is a custom python wrapper over `docker_cli` that mimics the interface of `docker_py`. 
    - The file has 3 additional classes, `Container`, `Compose`, and `Image`.
- Removed a lot of functions from existing files (all calls to subprocess) and docker errors. Migrated them to the `docker_client.py` and replaced errors with our new errors.
- `teardown` function in cli has the most amount of changes as handling of `--all` flag is moved to `engine.teardown` instead as we no longer need to filter through all the projects in docker and parse which ones belong to Tesseract as our  `DockerWrapper` class already knows all the projects that are Tesseracts.
- Renamed `_display_container_meta` to `_display_project_meta` as I think it's more fitting to the function since we are listing the metadata of all the containers in a project, aka the project's metadata.
- Modified some unit tests to fit the new mocked class.

BREAKING CHANGE: Tesseract Core does not clear build caches anymore. That means, the option `--keep-build-cache` and environment variable `TESSERACT_KEEP_BUILD_CACHE` are now removed, and users are advised to clean up Docker resources manually as needed (e.g. via `docker system prune`).

#### Testing done
Unit tests + Manually tested commands:
`serve` `build` `ps` `list` `teardown` `teardown --all` `apidoc`

#### License

- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://pasteurlabs.github.io/tesseract/LICENSE).
- [x] I sign the Developer Certificate of Origin below by adding my name and email address to the `Signed-off-by` line.

<details>
<summary><b>Developer Certificate of Origin</b></summary>

```text
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

Signed-off-by: [Angela Ko] <angela.ko@simulation.science>

dionhaefner commented on tesseract_core/sdk/cli.py:
```diff
@@ -296,9 +290,10 @@ def build_image(
-    except docker.errors.BuildError as e:
-        raise UserError(f"Error building Tesseract: {e}") from e
-    except docker.errors.APIError as e:
+    except RuntimeError as e:
+        # If "build" in error message:
+        if "build" in str(e):
+            raise UserError(f"Error building Tesseract: {e}") from e
```
Can we use a custom exception instead? I.e., define `DockerCLIException` or so that we raise whenever a command fails.

dionhaefner commented on tesseract_core/sdk/docker_cli_wrapper.py:
```diff
@@ -0,0 +1,475 @@
+logger = logging.getLogger("tesseract")
+
+
+class DockerWrapper:
```
Thinking of a better name for this class. `CLIDockerClient`? To me it's more than a wrapper, and closely modeled after `DockerClient`, so the name should reflect that.

dionhaefner commented on tesseract_core/sdk/engine.py:
```diff
@@ -591,69 +499,8 @@ def serve(
-        logger.error(ex.stderr.decode())
-        return False
+        project_id = docker_wrapper.docker_compose_up(compose_file.name, project_name)
+        return project_id, docker_wrapper
```
Why return the client?

dionhaefner replied:
In that case it could also make sense to use a singleton object (like a global instance of the client) instead.

dionhaefner commented on tests/sdk_tests/test_engine.py:
```diff
@@ -54,10 +49,10 @@ def test_build_image(
         assert command in caplog.text
     else:
-        assert got.attrs == mocked_docker.images.get(image_name).attrs
+        assert got.attrs == mocked_docker.get_image(image_name).attrs
```
I think this PR would be most impactful (as in, increase re-usability of the CLI client) if we can minimize the diff in tests – if both objects quack ~ the same, we can switch back and forth between them, here and in other projects. Do you think it's feasible to adopt the `DockerClient` interface here, with `client.images.get` and `client.containers.get` instead of `client.get_image` / `client.get_container`?

angela-ko commented:
It's ready.. @dionhaefner 

i updated the PR body with the new information, and 1 failing unit test that I added myself that passes locally but fails on github ci for some reason... Has to do with container logs which is another questionable point that I'm not sure if I'm handling correctly or not. 

I wonder if it has to do with the fact that because I'm running with detached containers, the containers automatically stop when the dummy command (echo) is done running. To address this I added a sleep infinitely command, but maybe that command doesn't work in github-ci so the containers are stopping early?? Idk.. That's my current theory. 

Everything else is good to go.. I could probably add some more unit tests to test the image name handling since I'm also unsure about that (since a different part of the code mentioned that image names could have repository url in it), but I'll do that tomorrow. 

I also added a bunch of functions that I didnt have before like `containers.container.run` `containers.container.log` `containers.container.remove` etc, because the old version of the code just ran with the `--rm` flag set to true to automatically handle removal, however if we want to have access to the container logs, we need to run with detach to get the container id and the logs from the container... The whole container set up was what ended up adding so much more time.... 

Let me know your thoughts for now.

xalelax commented:
I had a quick read and it looks fine, really nice work @angela-ko ! I will give it a more in-depth review on Monday -- and I don't think we should press the button for this today 🙂

dionhaefner commented on tests/conftest.py:
```diff
@@ -82,6 +83,22 @@ def unit_tesseract_path(request) -> Path:
 
+@pytest.fixture()
+def dummy_docker_file(tmpdir):
+    """Create a dummy Dockerfile for testing."""
```
Please add a description why this is needed (esp. the infinite sleep).

dionhaefner commented on tests/endtoend_tests/test_docker_cli_wrapper.py:
How come you chose not to test against the `docker-py` client?

angela-ko replied:
I thought that the entire point of this PR was to remove our dependency on docker-py. If we were to test against docker-py, that would mean that we would have docker-py as a dependency for testing which would lead to devs encountering the errors like the EXPORT_HOST bug again... which is kind of frustrating just for the sake of testing

dionhaefner replied:
I don't really care whether devs needs to jump through a few hoops to set up an environment, that's kind of expected. Testing against the official client provides us with extra confidence when it comes to API compatibility – i.e., can you invoke both objects with the same arguments and get identical results?

(That's not a must at this point, happy to discuss more.)

dionhaefner commented on tests/sdk_tests/test_engine.py:
```diff
@@ -55,8 +50,11 @@ def test_build_image(
-        command = f"docker buildx build --load --tag {image_name} --file {dockerfile_path} {tmpdir}"
-        assert command in caplog.text
+        # Split into 2 parts because label is timestamp based
+        command1 = f"docker buildx build --load --tag {image_name} --label"
+        command2 = f"--file {dockerfile_path} {tmpdir}"
```
Should the `--label` part actually be printed by this command? And if so, should we also print the command to clear the cache after?

angela-ko replied:
I supposed that is optional... 

And if someone is trying to manually run docker build.. surely they know how to clean up after themselves without us printing out the prune command for them.. right?

dionhaefner commented:
> I also added a bunch of functions that I didnt have before like containers.container.run containers.container.log containers.container.remove etc, because the old version of the code just ran with the --rm flag set to true to automatically handle removal, however if we want to have access to the container logs, we need to run with detach to get the container id and the logs from the container... The whole container set up was what ended up adding so much more time....

In which cases do we need to grab the logs via `docker logs`, vs. just capturing what's written to `stdout` / `stderr`?

angela-ko commented on tests/conftest.py:
```diff
@@ -180,9 +196,11 @@ def dummy_image_name(docker_client):
-                pass
+                docker_client.images.remove(image_name)
+            except RuntimeError as ex:
+                # If image is not found is in the exception error string, pass
+                if "Cannot remove image" in str(ex).lower():
+                    pass
```
oh, I suppose I should just except ImageNotFound Errors only like the original code so that everything else gets raised

angela-ko commented on tesseract_core/sdk/docker_cli_wrapper.py:
As you can see from the unit tests and me reverting most of the changes I made in the previous pr, it has the exact same function names and signatures  and nested objects as the docker-py client, in fact, the unit tests can all pass if you replace docker-client with the docker-py client again because I didn't touch any of the unit tests.

The only modification I changed to a unit test was the example unit test that calls `docker.containers.container.run(...serve)` because that one is 1) outdated 2) calls the docker run command directly with a `ports` argument that was difficult to implement and also unnecessary since serving with `tesseract run` is deprecated anyways). 

It is a **subset of the docker-py** though since I only implemented the functions & arguments that we actually use, so stuff like `containers.attach_socket` `containers.commit` etc aren't implemented, and the additional kwargs in `run` I mentioned above that we don't use.

dionhaefner replied:
That sounds great, but is it really a strict subset? For example I see the method `buildx` which has some Tesseract-specific logic (like SSH forward an whatnot)?

angela-ko commented:
> > I also added a bunch of functions that I didnt have before like containers.container.run containers.container.log containers.container.remove etc, because the old version of the code just ran with the --rm flag set to true to automatically handle removal, however if we want to have access to the container logs, we need to run with detach to get the container id and the logs from the container... The whole container set up was what ended up adding so much more time....
> 
> In which cases do we need to grab the logs via `docker logs`, vs. just capturing what's written to `stdout` / `stderr`?


Anything that's run in the container doesn't get printed to stdout when we run in detached form, stdout only prints the container id so if we want to retrieve the logs that occurred in the run command, we would need to call `docker logs`. 

On the same topic, we want to run with detached because that is the only way to determine what the container id of the container we just spawned is if we wish to retrieve the container obj, and for max compatibility with the existing unit tests logic and docker-py api, we need to be able to return the container obj for processing (this actually was the part that took the longest to figure out).

dionhaefner commented:
@angela-ko I had a closer look and fixed the issue with logging. I'll fix the remaining tests and smaller issues. 

In the meantime, please apply the following adjustments:

- [ ] Add extensive documentation (docstrings) outlining the scope + design philosophy of your CLI Docker client, essentially summarizing the discussion above.
- [ ] Add unit tests against `docker-py` to increase confidence in the operational readiness of your implementation.

angela-ko commented on tesseract_core/sdk/docker_cli_wrapper.py:
```diff
@@ -0,0 +1,613 @@
+        self.images = self.Images()
+        self.compose = self.Compose(self)
+
+    class Images:
```
This is because the docker-py classes are lower case for some reason, so our class is Images and I initialize them in the client as lower cased `.images.` which allows us to do stuff like `client.containers.container.run`

angela-ko commented on tests/conftest.py:
```diff
@@ -298,8 +298,15 @@ def list() -> list[docker_cli_wrapper.CLIDockerClient.Images.Image]:
-            def buildx(*args, **kwargs) -> str:
-                return docker_cli_wrapper.CLIDockerClient.Images.buildx(*args, **kwargs)
+            def buildx(
+                *args, print_and_exit=False, **kwargs
```
oh didnt know you could do that so i had to just call the real out to get the print and exit test! very cool

angela-ko commented on tesseract_core/sdk/docker_client.py:
```diff
@@ -166,7 +169,7 @@ def buildx(
                 raise CLIDockerClient.Errors.BuildError(logs)
 
-            return tag
+            return self.get(tag)
```
I originally had it like this but changed it to just returning the tag instead of getting the image in this line since I think thats how the original code had it as / for debugging purposes to better isolate the issue (forgot what the exact bug was, possibly for better error handling too)

dionhaefner replied:
Let's keep it this way for parity with `docker-py`.

angela-ko commented on tesseract_core/sdk/docker_client.py:
```diff
@@ -424,7 +430,7 @@ def _update_containers(self) -> None:
                 self.containers = {}
                 result = subprocess.run(
-                    ["docker", "ps", "-q"],  # List only container IDs
+                    ["docker", "ps", "-q", "--all"],  # List only container IDs
```
I thought we only want to show running containers if we call tesseract ps?

angela-ko commented on tests/sdk_tests/test_engine.py:
```diff
@@ -246,15 +243,17 @@ def test_serve_tesseracts(mocked_docker):
 
 
-def test_needs_docker(mocked_docker):
+def test_needs_docker(mocker):
```
should we be passing in mocked_docker in here because engine.needs_docker implicitly calls docker_client, so we would need to get it to be mocked here? 

similarily, the patch is only patching the real version's info and not the mocked_docker's info. 

I noticed that we are failing the `test_needs_docker` and this is what I ended up doing to get the test to pass:
- calling real docker.info() in my mock info() func and patching the real docker info()'s output to raise the error

(this test i always struggled with though so maybe that's not the best solution here)

dionhaefner replied:
Fixed, thanks.

(Solution is to use the mocked docker, and monkeypatch that object's `info` method.)

dionhaefner commented on tesseract_core/sdk/docker_client.py:
```diff
@@ -0,0 +1,616 @@
+            # NOTE: Do this before error checking to ensure we always prune the cache
+            if not keep_build_cache:
+                try:
+                    prune_cmd = [
+                        "docker",
+                        "builder",
+                        "prune",
+                        "-a",
+                        "-f",
+                        "--filter",
+                        f"label={label}",
```
This doesn't seem to work, for me it always prints `0B`. See also here: https://github.com/docker/cli/issues/2450#issuecomment-754033345

Should we bring back the `until:` logic?

angela-ko replied:
I see -- I thought it was working since the unit test that was failing because of insufficient space started passing after I added that, but it seems the entire filter flagged was broken in 2021 but only the `until` portion got fixed in https://github.com/docker/buildx/pull/1252 ... 

https://github.com/docker/buildx/blob/master/commands/prune.go#L191-L197

and from this it seems like they don't even support labels anymore period.. so I'm good with going back to the old until logic, but shouldn't the pruning be until the build completes timestamp instead of the start time like we had before? (Was always confused about that part)

xalelax commented:
I had a brief look at the code + threads, and don't have much to add to this 👍 gj

angela-ko commented:
Fixed issue with a singular docker image with multiple tags and handling of short id. 

Current unit tests fail only on github ci not locally so sorry about the incoming print statement spamming...

angela-ko commented:
Ok - unit tests fixed except for the permission error tried changing it to 777 but it still doesn't work, even though it works locally... 

Added a flag to view non-tesseract images / containers with the client as well.

angela-ko commented:
Dangling images (i.e. images with no tags) are not listed by the client -- not sure if there is a use case where we would want to list them? Because if we do, we would need some additional handling regarding image names.

angela-ko commented:
@dionhaefner tests passing i think we're in a pretty good state now

dionhaefner commented on tests/endtoend_tests/test_endtoend.py:
```diff
@@ -234,6 +231,10 @@ def test_tesseract_teardown_multiple(built_image_name, tear_all):
             args.extend(project_ids)
 
+        if len(args) < 2:
+            # No projects to teardown
+            pass
```
This is effectively dead code (`pass` just means continue executing)

dionhaefner reviewed (commented):
Awesome, thanks @angela-ko! Getting real close now, let's finish up the final few things and finally ship it.

angela-ko commented on tesseract_core/sdk/docker_client.py:
```diff
@@ -0,0 +1,831 @@
+                    "docker",
+                    "builder",
+                    "prune",
+                    "-a",
```
I think there might have been a misunderstanding of the old behavior which is why I just went with -a since it effectively does the same thing. 

From the comments of our old logic, it seemed like we wanted to track our start time stamp of this job and prune cache starting from the timestamp; however, the prune filter actually prunes everything up until that timestamp so the actual build cache we just created doesn't get pruned. 

https://docs.docker.com/reference/cli/docker/system/prune/

system prune which I think has the same logic as builder prune -- 

> until (<timestamp>) - only remove containers, images, and networks created before given timestamp

So pruning everything up until the timestamp of the job we just created seems to be counterintuitive as it seemed like we wanted to only prune the job we created. Since labels doesn't work, I just swapped to `-a` since it effectively is doing the same thing as pruning everything up until the start time of our build command + including the one we just built. 

Not sure if that makes sense! Let me know your thoughts

dionhaefner replied:
I see. Looks like you're right.

Since people were on the fence regarding the utility of cache pruning in the first place, let's remove the feature.

angela-ko commented on tesseract_core/sdk/docker_client.py:
```diff
@@ -0,0 +1,831 @@
+list_ = list
+
+
+class Image:
```
very cool

dionhaefner commented on tesseract_core/sdk/docker_client.py:
```diff
@@ -0,0 +1,831 @@
+            raise ImageNotFound(f"Cannot remove image {image}: {res.stderr}")
+
+    @staticmethod
+    def buildx(
```
Done, please have a look (no drastic changes, just imo better separation of concerns).
